treeherder/perf/models.py

```diff
@@ -227,9 +229,50 @@ def _compute_ideal_chunk_size(self, max_timestamp, max_chunk_size):
         return len(older_ids) or max_chunk_size
 
 
-class PerformanceDatum(models.Model):
-    objects = PerformanceDatumManager()
+class Counter(models.Model):
+    """
+    Used for tracking the blocks of ids severed to processes
+    Supports dense, globally unique, ids
+    """
+
+    model = models.CharField(max_length=80, primary_key=True)
+    chunk_size = models.IntegerField(null=False, default=100)
+    count = models.BigIntegerField(null=False)
+
+
+class AutoIncrementMixin:
+    COUNTERS = {}
+    COUNTERS_LOCK = Lock()
+
+    def next_id(self):
+        """
+        REQUIRES 'isolation_level': 'serializable',
+        :return:  A unique integer key for given model
+        """
+        model = type(self)
+        model_name = model.__name__
+        with AutoIncrementMixin.COUNTERS_LOCK:
+            count, maxx = AutoIncrementMixin.COUNTERS.get(model_name, (1, 0))
+            if count >= maxx:
+                # we require a new block of ids
+                with transaction.atomic():
+                    db_counter = Counter.objects.filter(model=model_name).first()
+                    if not db_counter:
+                        # only happens once
+                        start = model.objects.aggregate(id=Max("id"))['id'] or 0
+                        db_counter = Counter.objects.create(
+                            model=model_name, chunk_size=100, count=start + 1
+                        )
+                    count = db_counter.count
+                    maxx = db_counter.count = count + db_counter.chunk_size
```

Comment:
@armenzg how is Treeherder configured to handle requests, in terms of threads/processes?
Is it using thread/process pools? Is it spawning new processes per request, then recycles them?

If the latter case is true, then it's possible that each process grabs a chunk of size 100 from the database, uses less ids than that, then it terminates. The left ids will remain unused and unavailable, according to the `Counter` table.
@klahnakoski am I right about this? Are we expecting to waste some of the provided ids?

Reply:
@ionutgoldan Yes, you got it correct. There will be holes in the id sequence when a process ends.  The maximum size of the holes is adjustable using the `chunk_size` which can be set dynamically in the database at any time.

Reply:
I don't know Ionut. I know less about databases than people think.